File: 0.0/Shareable/Shareable/STransaction.cs
using System;
#nullable enable

namespace Shareable
{
    public class STransaction :SDatabase 
    {
        // uids above this number are for uncommitted objects in this transaction
        // Note: uncommitted objects of any type are added to tr.objects 
        public static readonly long _uid = 0x4000000000000000;
        public readonly long uid;
        public readonly bool autoCommit;
        public readonly SDatabase rollback;
        public readonly SDict<long, bool> readConstraints;
        internal override SDatabase _Rollback => rollback;
        protected override bool Committed => false;
        public STransaction(SDatabase d,bool auto) :base(d)
        {
            autoCommit = auto;
            rollback = d._Rollback;
            uid = _uid;
            readConstraints = SDict<long, bool>.Empty;
        }
        /// <summary>
        /// Some other set of updates to existing (and maybe named) objects 
        /// </summary>
        /// <param name="tr"></param>
        /// <param name="obs"></param>
        /// <param name="nms"></param>
        protected STransaction(STransaction tr, SDict<long,SDbObject> obs,SDict<string,SDbObject> nms,long c)
            : base(tr, obs, nms, c)
        {
            autoCommit = tr.autoCommit;
            rollback = tr.rollback;
            uid = tr.uid + 1;
            readConstraints = tr.readConstraints;
        }
        protected STransaction(STransaction tr,long u) :base(tr)
        {
            autoCommit = tr.autoCommit;
            rollback = tr.rollback;
            uid = tr.uid;
            readConstraints = tr.readConstraints + (u, true);
        }
        public static STransaction operator+(STransaction tr,long uid)
        {
            return new STransaction(tr, uid);
        }
        protected override SDatabase New(SDict<long, SDbObject> o, SDict<string, SDbObject> ns, long c)
        {
            return new STransaction(this,o, ns, c);
        }
        protected override Serialisable _Get(long pos)
        {
            if (pos < 0 || pos >= _uid)
                return objects[pos];
            return base._Get(pos);
        }
        public SDatabase Commit()
        {
            var f = dbfiles[name];
            SDatabase db = databases[name];
            var rdr = new Reader(f, curpos);
            var since = rdr.GetAll(db,db.curpos);
            for (var i = 0; i < since.Length; i++)
            {
                if (since[i].Check(readConstraints))
                    throw new Exception("Transaction conflict with read");
                for (var b = objects.PositionAt(_uid); b != null; b = b.Next())
                    if (since[i].Conflicts(b.Value.Item2))
                        throw new Exception("Transaction conflict on " + b.Value);
            }
            lock (f)
            {
                since = rdr.GetAll(this, f.length);
                for (var i = 0; i < since.Length; i++)
                {
                    if (since[i].Check(readConstraints))
                        throw new Exception("Transaction conflict with read");
                    for (var b = objects.PositionAt(_uid); b != null; b = b.Next())
                        if (since[i].Conflicts(b.Value.Item2))
                            throw new Exception("Transaction conflict on " + b.Value);
                }
                db = f.Commit(db,this);
                f.CommitDone();
            }
            Install(db);
            return db;
        }
        /// <summary>
        /// We will single-quote transaction-local uids
        /// </summary>
        /// <returns>a more readable version of the uid</returns>
        internal static string Uid(long uid)
        {
            return SDbObject._Uid(uid);
        }
        public override STransaction Transact(bool auto=true)
        {
            return this; // ignore the parameter
        }
        public override SDatabase Rollback()
        {
            return rollback;
        }
    }
}

File: 0.1/ShareableJ/src/org/shareabledata/STransaction.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package org.shareabledata;
/**
 *
 * @author Malcolm
 */
public class STransaction extends SDatabase {
        // uids above this number are for uncommitted objects
        public static final long _uid = 0x40000000;
        public final long uid;
        public final boolean autoCommit;
        public final SDatabase rollback;
        public final SDict<Long,Boolean> readConstraints;
        @Override
        protected boolean getCommitted(){
                return false;
        }
        @Override
        SDatabase getRollback() {
            return rollback;
        }
        public STransaction(SDatabase d, ReaderBase rdr,boolean auto)
        {
            super(d);
            uid = _uid;
            autoCommit = auto;
            rollback = d.getRollback();
            readConstraints = null;
            rdr.db = this;
        }
        private STransaction(STransaction tr,SDict<Long,SDbObject>obs,SRole r,long c) throws Exception
        {
            super(tr,obs,r,c);
            uid =  tr.uid+1;
            autoCommit = tr.autoCommit;
            rollback = tr.rollback;
            readConstraints = tr.readConstraints;
        }
        protected STransaction(STransaction tr,long u)
        {
            super(tr);
            autoCommit = tr.autoCommit;
            rollback = tr.rollback;
            uid = tr.uid;
            readConstraints = (tr.readConstraints==null)?new SDict<>(u,true):
                    tr.readConstraints.Add(u, true);
        }
        public STransaction(STransaction tr,SRole r)
        {
            super(tr,tr.objects,r,tr.curpos);
            autoCommit = tr.autoCommit;
            rollback = tr.rollback;
            uid = tr.uid;
            readConstraints = tr.readConstraints; 
        }
        // Add a readConstraint : NB creates a new STransaction
        public STransaction Add(long u)
        {
            return new STransaction(this,u);
        }
        public STransaction Add(String s)
        {
            return (STransaction)New(objects,role.Add(uid,s),curpos);
        }
        public STransaction Add(long u,SDbObject ob)
        {
            return (STransaction)New(objects.Add(u,ob),role,curpos);
        }
        @Override
        public Serialisable _Get(long pos) {
            if (pos<0 || pos>=_uid)
                return objects.Lookup(pos);
            return super._Get(pos);
        }
        @Override
        protected SDatabase New(SDict<Long, SDbObject> o, SRole ro, long c)
        {
            STransaction r;
            try {
                r = new STransaction(this,o, ro, c);
            } catch (Exception e){ r = this; }
            return r;
        }
        @Override
        public String Name(long u)
        {
            return role.uids.get(u);
        }
        /// <summary>
        /// If there are concurrent transactions there will be more code here.
        /// </summary>
        /// <returns>the steps as modified by the commit process</returns>
        public SDatabase Commit() throws Exception
        {
            SDatabase db = databases.Lookup(name);
            var f = new Writer(dbfiles.get(name));
            var rdr = new Reader(db,curpos);
            var tb = objects.PositionAt(_uid); // start of the work we want to commit
            var since = rdr.GetAll(f.length());
            for (SDbObject since1 : since) {
                if (since1.Check(readConstraints))
                    throw new Exception("Transaction conflict with read");
                for (org.shareabledata.Bookmark<org.shareabledata.SSlot<java.lang.Long, org.shareabledata.SDbObject>> b = objects.PositionAt(_uid); b != null; b = b.Next()) {
                    if (since1.Conflicts(db,this,b.getValue().val)) {
                        throw new Exception("Transaction conflict on " + b.getValue());
                    }
                }
            }
            synchronized (f)
            {
                since = rdr.GetAll(f.length());
                for (SDbObject since1 : since) {
                    if (since1.Check(readConstraints))
                        throw new Exception("Transaction conflict with read");
                    for (org.shareabledata.Bookmark<org.shareabledata.SSlot<java.lang.Long, org.shareabledata.SDbObject>> b = objects.PositionAt(_uid); b != null; b= b.Next()) {
                        if (since1.Conflicts(db,this,b.getValue().val)) {
                            throw new Exception("Transaction conflict on " + b.getValue());
                        }
                    }
                }
                db = f.Commit(db,this);
                f.CommitDone();
            }
            Install(db);
            return db;
        }
        @Override
        public STransaction Transact(ReaderBase rdr,boolean auto)
        {
            rdr.db = this;
            return this; // ignore the parameter
        }
        @Override
        public SDatabase Rollback()
        {
            return rollback;
        }
    }

